dev/fattest.simplicity/src/componenttest/topology/impl/LibertyServer.java

```diff
@@ -3474,6 +3486,146 @@ protected void checkLogsForErrorsAndWarnings(List<String> failuresRegExps, Strin
+                                                     "SlowAppStartupHealthCheck", //io.openliberty.microprofile.health31.fat.SlowAppStartupHealthCheckTest
+                                                     "FATServer", //org.eclipse.microprofile.graphql.tck.GraphQLTckPackageTest
+    };
+    private static final Set<String> EXEMPT_SERVERS_SET = new HashSet<String>(Arrays.asList(EXEMPT_SERVERS));
+
+    /**
+     * After the server has shutdown, check that the features used at runtime matched those expected by any FeatureReplacementAction which may have been active.
+     *
+     * @throws Exception
+     */
+    protected void checkServerRepeatFeatures() throws Exception {
+        String method = "checkServerRepeatFeatures";
+
+        RepeatTestAction action = RepeatTestFilter.getMostRecentRepeatAction();
+        if (action != null && action instanceof FeatureReplacementAction) {
+            FeatureReplacementAction featureReplacementAction = (FeatureReplacementAction) action;
+
+            //only check the features if this server was included in the FeatureReplacementAction
+            String serverName = getServerName();
+
+            Set<String> servers = featureReplacementAction.getServers();
+            //only check the features if the FeatureReplacementAction applies to this server
+            if (!servers.isEmpty() &&
+                !servers.contains(FeatureReplacementAction.NO_SERVERS) &&
+                (servers.contains(serverName) || servers.contains(FeatureReplacementAction.ALL_SERVERS))) {
+
+                Set<String> expectedFeatures = new HashSet<>(featureReplacementAction.getAddFeatures()); //the expected features, if present
+                expectedFeatures.addAll(featureReplacementAction.getAlwaysAddFeatures());
+                Set<String> installedFeatures = getInstalledFeatures(); //the features actually installed at runtime
+
+                //compare each installed feature to the expected ones
+                for (String installedFeature : installedFeatures) {
+                    //ignore versionless features
+                    int dash = installedFeature.indexOf("-");
+                    if (dash > -1) {
+                        String versionlessInstalledFeature = removeFeatureVersion(installedFeature);
+                        for (String replacementFeature : expectedFeatures) {
+                            String versionlessReplacementFeature = removeFeatureVersion(replacementFeature);
+                            if (versionlessReplacementFeature.equalsIgnoreCase(versionlessInstalledFeature)) {
+                                //if the base feature name matches but the version does not, throw an exception.
+                                if (!replacementFeature.equalsIgnoreCase(installedFeature)) {
+                                    String message = "Runtime feature was not of the expected version for repeat action (" + serverName + ", " + action.getID() + "). "
```

Comment:
```suggestion
                                    String message = "Runtime feature was not of the expected version for repeat action (server: " + serverName + ", action id: " + action.getID() + "). "
```

Looking at the results of your build failure, it's not always obvious that the server name is a server name.

Reply:
Done